alpine/ParticleContainer.hpp

```diff
@@ -9,10 +9,11 @@ template <typename T, unsigned Dim = 3>
 class ParticleContainer : public ippl::ParticleBase<ippl::ParticleSpatialLayout<T, Dim>>{
```

Comment:
If the particle container class derives from `ParticleBase`, why is the particle container a template parameter of the PIC manager? Why isn't the member just a pointer to `ParticleBase`?

Reply:
Actually, I tried it before. The issue is that then `pcontainer_m` of `PicManager` in src/Manager/PicManager.h would be a pointer of type `ippl::ParticleBase<ippl::ParticleSpatialLayout<T, Dim>` which doesn't have P,q,E as public variables. Then, the instance of ParticleContainer in the alpine application cannot be used to set `pcontainer_m` of `PicManager`, since the public members (P,q,E) are not declared in the base class.


Maybe there's a way to do it nicely, but I when I tried,  got errors like this

```
error: ‘using element_type = class ippl::ParticleBase<ippl::ParticleSpatialLayout<double, 3, ippl::UniformCartesian<double, 3> > >’ {aka ‘class ippl::ParticleBase<ippl::ParticleSpatialLayout<double, 3, ippl::UniformCartesian<double, 3> > >’} has no member named ‘P’
  266 |         view_type* P = &(pcontainer_m->P.getView());
      |                                        ^
/psi/home/sadr_m/Codes/ippl_fresh/ippl/alpine/LandauDampingManager.h:280:23: error: ‘using element_type = class ippl::ParticleBase<ippl::ParticleSpatialLayout<double, 3, ippl::UniformCartesian<double, 3> > >’ {aka ‘class ippl::ParticleBase<ippl::ParticleSpatialLayout<double, 3, ippl::UniformCartesian<double, 3> > >’} has no member named ‘q’
  280 |         pcontainer_m->q = Q_m/totalP;
      |                       ^
/psi/home/sadr_m/Codes/ippl_fresh/ippl/alpine/LandauDampingManager.h: In member function ‘void LandauDampingManager::LeapFrogStep()’:
/psi/home/sadr_m/Codes/ippl_fresh/ippl/alpine/LandauDampingManager.h:303:51: error: conversion from ‘shared_ptr<ippl::ParticleBase<ippl::ParticleSpatialLayout<double, 3, ippl::UniformCartesian<double, 3> > >>’ to non-scalar type ‘shared_ptr<ParticleContainer<double, 3>>’ requested
  303 |         std::shared_ptr<ParticleContainer_t> pc = pcontainer_m;
```

Reply:
I think this is largely an issue of what we want the PIC manager to represent. As I mentioned in the Matrix channel, the PIC manager in IPPL core currently does nothing. For this reason, all of its members are also unnecessary restrictions (i.e. why does PIC manager care what data type the particle container is? It doesn't do anything with the particles). ALPINE is forced to propagate a data type into IPPL core in order to use that data type at application level.

Instead, the PIC manager could take the particle layout as its template parameter (as the original `ChargedParticles` did) and have a `std::vector<ParticleBase<PLayout>*>` member. This is more semantically meaningful, as a simulation does not necessarily have just one particle type, nor should the generic PIC manager depend on the exact particle properties. The vector contains pointers because the PIC manager base class does not own the particles. It merely acts on them (or will, eventually). Instead, the particles belong to the subclass (i.e. `class AlpineManager : public PICManager`), where the particle bunch has a data type that extends `ParticleBase` in order to describe a specific particle type.

In the following skeleton code, IPPL's core PIC manager is unaware of the particle properties. At the application level, the user defines their own particle type and extends the manager class to create a new manager class specific to a simulation containing bunches of that particle. The user's manager class is then free to access the specific particle properties as needed and the core PIC manager does not need access to the user's particle type.

```cpp
template <typename PLayout>
class PICManager {
  std::vector<ParticleBase<PLayout>*> particles;
};

class TachyonBunch : public ParticleBase<...> { ... };

template <...>
class FTLManager : public PICManager<ParticleSpatialLayout<...>> {
  TachyonBunch tachyons;

  FTLManager() : tachyons(...) { this->particles.push_back(&tachyons); }
};
```

Reply:
I disagree, the PICManager should not take the particle layout as a template argument. I see the PICManager as a class that can perform PIC simulations, i.e. it has a stepper, particle container (or more than one), field solver elements and maybe a particle load balancer.

Reply:
I otherwise agree with above example.